dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorOpsBarrier.java

```diff
@@ -0,0 +1,73 @@
+package com.dat3m.dartagnan.parsers.program.visitors.spirv;
+
+import com.dat3m.dartagnan.exception.ParsingException;
+import com.dat3m.dartagnan.expression.Expression;
+import com.dat3m.dartagnan.expression.ExpressionFactory;
+import com.dat3m.dartagnan.expression.type.IntegerType;
+import com.dat3m.dartagnan.expression.type.TypeFactory;
+import com.dat3m.dartagnan.parsers.SpirvBaseVisitor;
+import com.dat3m.dartagnan.parsers.SpirvParser;
+import com.dat3m.dartagnan.parsers.program.visitors.spirv.helpers.HelperTags;
+import com.dat3m.dartagnan.parsers.program.visitors.spirv.builders.ProgramBuilder;
+import com.dat3m.dartagnan.program.event.Event;
+import com.dat3m.dartagnan.program.event.EventFactory;
+import com.dat3m.dartagnan.program.event.Tag;
+
+import java.util.Set;
+
+public class VisitorOpsBarrier extends SpirvBaseVisitor<Event> {
+
+    private static final ExpressionFactory expressions = ExpressionFactory.getInstance();
+    private final IntegerType archType = TypeFactory.getInstance().getArchType();
+    private final ProgramBuilder builder;
+    private int nextBarrierId = 0;
+
+    public VisitorOpsBarrier(ProgramBuilder builder) {
+        this.builder = builder;
+    }
+
+    @Override
+    public Event visitOpControlBarrier(SpirvParser.OpControlBarrierContext ctx) {
+        if (!ctx.execution().getText().equals(ctx.memory().getText())) {
+            throw new ParsingException("Unequal scopes in OpControlBarrier are not supported");
+        }
+        Expression barrierId = expressions.makeValue(nextBarrierId++, archType);
+        Event barrier = EventFactory.newControlBarrier("cbar", barrierId);
+        barrier.addTags(Tag.Spirv.CONTROL);
+        barrier.addTags(getScopeTag(ctx.execution().getText()));
+        Set<String> tags = getMemorySemanticsTags(ctx.semantics().getText());
+        if (Set.of(Tag.Spirv.RELAXED).equals(tags)) {
```

Comment:
In principle, a barrier can have avvis semantics but be relaxed. In this case it can have some memory semantics effect. But I am not sure if we will find it in a real code compiled from opencl.

Reply:
> In this case it can have some memory semantics effect. 

Do you mean that if the barrier is both relaxed and avvis, the FENCE tag should be kept?

Reply:
yes